urlc175f05cb8_mchllbrt_PermCode_tgz-pJ8-permlib_examples_UniversalSeparableJ8/src/permlab/ui/AnimatedPermFrame.java

```diff
@@ -0,0 +1,669 @@
+    private JMenuItem exportOption = new JMenuItem("Export current");
+    private JMenuItem exportAllAsTextOption = new JMenuItem("Export all in textbox");
+    private JMenuItem exportAllAsGIF = new JMenuItem("Export all as a GIF");
+
+    /**
+     * Creates the pop-up menu.
+     */
+    private void addMenu() {
+        popupMenu.add(exportOption);
+        exportOption.addActionListener(new ActionListener() {
+            @Override
+            public void actionPerformed(ActionEvent e) {
+                if (permIndex >= 0) {
+                    exportCurrent();
+                }
+            }
+        });
+        popupMenu.add(exportAllAsTextOption);
+        exportAllAsTextOption.addActionListener(new ActionListener() {
+            @Override
+            public void actionPerformed(ActionEvent e) {
+                TextFrame textDisplay = new TextFrame();
+                textDisplay.addText(perms);
+                textDisplay.setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE);
+                int x = displayPanel.getX();
+                textDisplay.setLocation(
+                        displayPanel.getX() + displayPanel.getWidth() + 10,
+                        displayPanel.getY());
+                textDisplay.pack();
+                textDisplay.setVisible(true);
+            }
+        });
+        popupMenu.add(exportAllAsGIF);
+        exportAllAsGIF.addActionListener(new ActionListener() {
+            @Override
+            public void actionPerformed(ActionEvent e) {
+                exportPermCollectionAsGIF(perms, displayPanel.getWidth(), displayPanel.getHeight());
+            }
+        });
+    }
+
+    /**
+     * Exports the entire collection as a GIF.
+     * NOTE: This GIF is set up so that it will display all permutations once
+     * without looping.
+     * 
+     * @param perms permutations to be exported
+     * @param width width of the GIF
+     * @param height height of the GIF
+     */
+    public void exportPermCollectionAsGIF(Permutation[] perms, int width, int height) {
+        BufferedImage image;
+        Graphics2D g;
+        File savingFile = new File("GIF.gif");
+        fileChooser.setSelectedFile(savingFile);
+        int returnVal = fileChooser.showSaveDialog(this);
+        if (returnVal == JFileChooser.APPROVE_OPTION) {
+            savingFile = fileChooser.getSelectedFile();
+            if (FileUtilities.canSaveFileHere(this, savingFile)) {
+                try {
+                    Permutation p = perms[0];
+                    image = new BufferedImage(width, height, BufferedImage.TYPE_INT_RGB);
+                    g = image.createGraphics();
+                    PaintUtilities.paint(g, p, width, height, false);
+
+                    ImageTypeSpecifier spec = new ImageTypeSpecifier(image);
+                    ImageWriter wr = ImageIO.getImageWriters(spec, "GIF").next();
+                    wr.setOutput(ImageIO.createImageOutputStream(savingFile));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource returned by the following method call: *createImageOutputStream*. Execution paths that do not contain closure statements were detected. Ensure the object returned by *createImageOutputStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).